percy-mirror-verifier/src/server.ts

```diff
@@ -0,0 +1,154 @@
+    const percyMainBuildId = webhooks.getBuildId(included);
+    console.log('Percy build id for main branch:', percyMainBuildId);
+    const githubPullNumber = webhooks.getPullNumber(included);
+    console.log(
+      'Pull request that was merged into this commit:',
+      githubPullNumber
+    );
+
+    const percyPullBuildId = await github.getPercyBuildId(githubPullNumber);
+    console.log('Percy build id for the pull request:', percyPullBuildId);
+
+    const prSnapshots = await percy.getSnapshots(percyPullBuildId);
+    const mainSnapshots = await percy.getSnapshots(percyMainBuildId);
+
+    if (prSnapshots.size === 1 && prSnapshots.has('Blank page')) {
+      console.log(
+        'Pull request',
+        githubPullNumber,
+        'was a docs-only change. Skipping...'
+      );
+      return;
+    }
+
+    // We ignore the Blank page.
+    mainSnapshots.delete('Blank page');
+
+    // Mirroring verification: the following must all be true to indicate a
+    // correct mirroring of the PR/main builds:
+    // * The number of snapshots on both the PR and main build is equal (after
+    //   we remove the "Blank page" snapshot)
+    // * The set of snapshot names are equal
+    // * All snapshots are approved
+    // * All snapshots must have the same approval reason (both are no_diff or
+    //   were both approved, manually/automatically)
+    // If any of these are false, this indicates a disparity between the PR and
+    // main builds.
+    if (
+      prSnapshots.size != mainSnapshots.size ||
+      [...prSnapshots.keys()].some(name => !mainSnapshots.has(name))
+    ) {
+      // TODO(danielrozenberg): write a comment on the PR (here and below).
+      console.error('PR/main disparity: not all snapshots exist in both');
+      return await github.postErrorComment(githubPullNumber);
+    }
+
+    if (
+      [...prSnapshots.values(), ...mainSnapshots.values()].some(
+        ({attributes}) => attributes['review-state'] !== 'approved'
+      )
+    ) {
+      console.error('PR/main disparity: not all snapshots are marked approved');
+      return await github.postErrorComment(githubPullNumber);
+    }
+
+    for (const [name, prSnapshot] of prSnapshots) {
+      const mainSnapshot = mainSnapshots.get(name);
+
+      const prApprovalReason = prSnapshot.attributes['review-state-reason'];
+      const mainApprovalReason = mainSnapshot.attributes['review-state-reason'];
+
+      if (
+        !(
+          (prApprovalReason === 'no_diffs' &&
+            mainApprovalReason === 'no_diffs') ||
+          (prApprovalReason === 'user_approved' &&
+            mainApprovalReason === 'auto_approved_branch')
+        )
+      ) {
+        console.error(
+          'PR/main disparity: not all snapshots have the same approval reason'
+        );
+        return await github.postErrorComment(githubPullNumber);
+      }
+    }
+
+    console.info('PR/main Percy builds are correctly mirrored!');
+  } catch (error) {
+    console.error(error);
```

Comment:
Suggestion: replace the contents of all `if` branches with `throw new Error('PR/main disparity: blah blah')` and do `console.error`/post error comment in the `catch`